logfire/_internal/config.py

```diff
@@ -1560,5 +1563,12 @@ def default_project_name():
     return sanitize_project_name(os.path.basename(os.getcwd()))
 
 
+def get_runtime_version() -> str:
```

Comment:
A doc string and some tests would be nice. Probably split it into a getter and a parser function so it can be tested.

Reply:
tests are already there

Reply:
with all possible inputs of `sys.implementation.version` e.g. pre-release versions? It'd be nice to see _explicitly_ what the expected inputs are.

Reply:
The issue is you can't instantiate the `sys.version_info` class to define custom values. So we could define a mock class acting like it but I don't know if it is worth it?

Reply:
I just want:

```python
def get_runtime_version(version_info: ?) -> str:
    if version_info.releaselevel == 'final' and not version_info.serial:
        return '.'.join(map(str, version_info[:3]))
    return '.'.join(map(str, version_info))  # pragma: no cover
```

And then tests along the lines of:

```python
def test_get_runtime_version_pre_release() -> None:
    version_info = sys.VersionInfo(???)
    version = get_runtime_version(version_info)
    assert version == ??
```

Maybe with parametrize

Reply:
Yes but you can't instantiate version infos, meaning we'd have to do something like:

```python
class MockVersionInfo(tuple[int, int, int, str, int]):
    @property
    def releaselevel(self) -> str: ...
    ...

get_runtime_version(MockVersionInfo(...))
```

Reply:
Is it an actual type?

```python
VersionInfo = tuple[int, int, int, str, int]

def get_runtime_version(version_info: VersionInfo) -> None:
    ...

Or worst case:

class SysVersionInfoProvider(Protocol):
    def __call__(self) -> VersionInfo: ...

def get_runtime_version(version_info: SysVersionInfoProvider) -> None:
    ...
```

Reply:
`sys.version_info` is an instance of the `version_info` class (confusingly of the same name, meaning my previous example doesn't make sense), but it can't be instantiated:

```python
type(sys.version_info)
#> <class 'sys.version_info'>
type(sys.version_info)(3, 9, 19, 'final', 0)
#> TypeError: cannot create 'sys.version_info' instances
```